File: backend/app/schemas/approval_workflow_templates.py
"""Request/response schemas for ``/api/approval-workflow-templates``.

PR #51 — reusable approval workflow blueprints. A template carries an
ordered list of step definitions; instantiation copies them into
concrete ``ApprovalStep`` rows on a new ``ApprovalWorkflowRun``. The
template itself is never mutated by an instantiation, and template edits
never touch in-flight runs.

Naming caution: ``AgreementTemplate`` (a document blueprint) is a
distinct concept. The instantiate request uses ``agreement_template_id``
to link a generated agreement, so it does not collide with the workflow
template path parameter.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

# ---------------------------------------------------------------------------
# Step schemas
# ---------------------------------------------------------------------------


class ApprovalWorkflowTemplateStepCreate(BaseModel):
    """Step payload accepted on template creation and step append.

    ``step_order`` is optional; if omitted the API assigns it as
    ``len(existing_steps) + 1`` for an append, or as the position in the
    create payload's ``steps`` array (1-indexed).
    """

    model_config = ConfigDict(extra="forbid")

    step_order: int | None = Field(default=None, ge=1)
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    approver_name: str | None = Field(default=None, max_length=255)
    approver_email: str | None = Field(default=None, max_length=255)
    assigned_to: uuid.UUID | None = None
    due_in_days: int | None = Field(default=None, ge=0, le=365)
    metadata_json: dict[str, Any] | None = None


class ApprovalWorkflowTemplateStepPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    step_order: int | None = Field(default=None, ge=1)
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    approver_name: str | None = Field(default=None, max_length=255)
    approver_email: str | None = Field(default=None, max_length=255)
    assigned_to: uuid.UUID | None = None
    due_in_days: int | None = Field(default=None, ge=0, le=365)
    metadata_json: dict[str, Any] | None = None


class ApprovalWorkflowTemplateStepResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    workflow_template_id: uuid.UUID
    step_order: int
    title: str
    description: str | None = None
    approver_name: str | None = None
    approver_email: str | None = None
    assigned_to: uuid.UUID | None = None
    due_in_days: int | None = None
    metadata_json: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Template schemas
# ---------------------------------------------------------------------------


# Free-form examples — kept narrow so the docstring suggests what to use
# without locking customers in. Backend stores it as an opaque string.
TEMPLATE_TYPE_EXAMPLES = (
    "legal_review",
    "finance_review",
    "procurement_review",
    "executive_approval",
    "general",
)


class ApprovalWorkflowTemplateCreate(BaseModel):
    """Body for ``POST /api/approval-workflow-templates``.

    At least one step is required: a template with no steps is not
    instantiable, so empty drafts are rejected at the API boundary.
    """

    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    template_type: str | None = Field(default=None, max_length=64)
    metadata_json: dict[str, Any] | None = None
    steps: list[ApprovalWorkflowTemplateStepCreate] = Field(min_length=1)


class ApprovalWorkflowTemplatePatch(BaseModel):
    """Body for ``PATCH /api/approval-workflow-templates/{template_id}``.

    Step edits go through the dedicated step endpoints; this body only
    updates template-level metadata. ``status`` is omitted on purpose:
    archiving uses the ``DELETE`` route (soft archive), and unarchiving
    is an explicit ``status`` field here.
    """

    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    template_type: str | None = Field(default=None, max_length=64)
    status: str | None = None
    metadata_json: dict[str, Any] | None = None


class ApprovalWorkflowTemplateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    name: str
    description: str | None = None
    template_type: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime
    created_by: uuid.UUID | None = None
    metadata_json: dict[str, Any] | None = None
    steps: list[ApprovalWorkflowTemplateStepResponse] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Instantiation
# ---------------------------------------------------------------------------


class CreateApprovalWorkflowFromTemplateRequest(BaseModel):
    """Body for ``POST /api/approval-workflow-templates/{template_id}/instantiate``.

    Mirrors the existing ad-hoc workflow create surface: at least one of
    ``request_id`` or ``contract_id`` is required, ``agreement_template_id``
    is the ``AgreementTemplate`` (document blueprint) link, deliberately
    spelled out to avoid colliding with the workflow template path
    parameter.
    """

    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=255)
    request_id: uuid.UUID | None = None
    contract_id: uuid.UUID | None = None
    agreement_template_id: uuid.UUID | None = None
    metadata_json: dict[str, Any] | None = None

    @model_validator(mode="after")
    def _require_request_or_contract(
        self,
    ) -> CreateApprovalWorkflowFromTemplateRequest:
        if self.request_id is None and self.contract_id is None:
            raise ValueError(
                "At least one of request_id or contract_id is required."
            )
        return self

File: frontend/src/types/approvalWorkflowTemplates.ts
/**
 * Types for the approval workflow templates surface (PR #51).
 *
 * Mirrors `backend/app/schemas/approval_workflow_templates.py`. A
 * workflow template is a reusable blueprint; instantiating one creates
 * a concrete `ApprovalWorkflowRun` plus `ApprovalStep` rows. The
 * template itself is never mutated by an instantiation.
 *
 * Naming caution: `AgreementTemplate` (a document blueprint) is a
 * different concept; the instantiate request carries it under
 * `agreement_template_id` so the two never collide on the wire.
 */

import type { ApprovalWorkflowRun } from "./approvalWorkflows";

export type ApprovalWorkflowTemplateStatus = "active" | "archived";

export interface ApprovalWorkflowTemplateStep {
  id: string;
  organization_id: string;
  workflow_template_id: string;
  step_order: number;
  title: string;
  description: string | null;
  approver_name: string | null;
  approver_email: string | null;
  assigned_to: string | null;
  /** Computed into a concrete `due_date = today + due_in_days` on instantiation. */
  due_in_days: number | null;
  metadata_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface ApprovalWorkflowTemplate {
  id: string;
  organization_id: string;
  name: string;
  description: string | null;
  template_type: string | null;
  status: ApprovalWorkflowTemplateStatus | string;
  created_at: string;
  updated_at: string;
  created_by: string | null;
  metadata_json: Record<string, unknown> | null;
  steps: ApprovalWorkflowTemplateStep[];
}

export interface ApprovalWorkflowTemplateStepCreate {
  step_order?: number | null;
  title: string;
  description?: string | null;
  approver_name?: string | null;
  approver_email?: string | null;
  assigned_to?: string | null;
  due_in_days?: number | null;
  metadata_json?: Record<string, unknown> | null;
}

export interface ApprovalWorkflowTemplateStepPatch {
  step_order?: number | null;
  title?: string | null;
  description?: string | null;
  approver_name?: string | null;
  approver_email?: string | null;
  assigned_to?: string | null;
  due_in_days?: number | null;
  metadata_json?: Record<string, unknown> | null;
}

export interface ApprovalWorkflowTemplateCreateRequest {
  name: string;
  description?: string | null;
  template_type?: string | null;
  metadata_json?: Record<string, unknown> | null;
  steps: ApprovalWorkflowTemplateStepCreate[];
}

export interface ApprovalWorkflowTemplatePatch {
  name?: string | null;
  description?: string | null;
  template_type?: string | null;
  status?: ApprovalWorkflowTemplateStatus | string | null;
  metadata_json?: Record<string, unknown> | null;
}

export interface ListApprovalWorkflowTemplateFilters {
  status?: string;
  template_type?: string;
  /** Defaults to false on the server. */
  include_archived?: boolean;
  /** Case-insensitive substring match on template name. */
  query?: string;
}

export interface CreateApprovalWorkflowFromTemplateRequest {
  name: string;
  request_id?: string | null;
  contract_id?: string | null;
  /** AgreementTemplate id (document blueprint), distinct from the workflow template. */
  agreement_template_id?: string | null;
  metadata_json?: Record<string, unknown> | null;
}

/** Instantiation returns a fully populated ApprovalWorkflowRun. */
export type CreateApprovalWorkflowFromTemplateResponse = ApprovalWorkflowRun;
